File: sdk/java/src/main/java/io/openthesis/Assert.java
package io.openthesis;

import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class Assert {
    private static final Set<String> declared = Collections.synchronizedSet(new HashSet<>());
    private static final Map<String, Boolean> everSinceArmed = new ConcurrentHashMap<>();

    public static void always(boolean condition, String message, Map<String, Object> details) {
        emit(condition, message, "always", true, details, 2);
    }

    public static void alwaysOrUnreachable(boolean condition, String message, Map<String, Object> details) {
        emit(condition, message, "always_or_unreachable", false, details, 2);
    }

    public static void sometimes(boolean condition, String message, Map<String, Object> details) {
        emit(condition, message, "sometimes", true, details, 2);
    }

    public static void reachable(String message, Map<String, Object> details) {
        emit(true, message, "reachable", true, details, 2);
    }

    public static void unreachable(String message, Map<String, Object> details) {
        emit(false, message, "unreachable", false, details, 2);
    }

    public static void everSince(boolean condition, String message, Map<String, Object> details) {
        String key = "ever_since:" + message;
        if (condition) {
            everSinceArmed.put(key, true);
        }
        boolean armed = everSinceArmed.containsKey(key);
        boolean effectiveCondition = condition || !armed;
        emit(effectiveCondition, message, "ever_since", true, details, 2);
    }

    public static void alwaysGreaterThan(long left, long right, String message, Map<String, Object> details) {
        Map<String, Object> d = details != null ? new java.util.HashMap<>(details) : new java.util.HashMap<>();
        d.put("left_value", left);
        d.put("right_value", right);
        emit(left > right, message, "always", true, d, 2);
    }

    public static void sometimesGreaterThan(long left, long right, String message, Map<String, Object> details) {
        Map<String, Object> d = details != null ? new java.util.HashMap<>(details) : new java.util.HashMap<>();
        d.put("left_value", left);
        d.put("right_value", right);
        emit(left > right, message, "sometimes", true, d, 2);
    }

    public static void alwaysEqual(long left, long right, String message, Map<String, Object> details) {
        Map<String, Object> d = details != null ? new java.util.HashMap<>(details) : new java.util.HashMap<>();
        d.put("left_value", left);
        d.put("right_value", right);
        emit(left == right, message, "always", true, d, 2);
    }

    public static void sometimesEqual(long left, long right, String message, Map<String, Object> details) {
        Map<String, Object> d = details != null ? new java.util.HashMap<>(details) : new java.util.HashMap<>();
        d.put("left_value", left);
        d.put("right_value", right);
        emit(left == right, message, "sometimes", true, d, 2);
    }

    public static void sometimesEach(String label, String key, Map<String, Object> details) {
        emit(true, label + ":" + key, "sometimes", true, details, 2);
    }

    public static void sometimesAll(String message, Map<String, Boolean> namedBools, Map<String, Object> details) {
        int satisfied = 0;
        int total = namedBools != null ? namedBools.size() : 0;
        if (namedBools != null) {
            for (boolean v : namedBools.values()) {
                if (v) satisfied++;
            }
        }
        boolean allTrue = total > 0 && satisfied == total;
        Map<String, Object> d = details != null ? new java.util.HashMap<>(details) : new java.util.HashMap<>();
        d.put("sub_goals", namedBools);
        d.put("satisfied_count", satisfied);
        d.put("total_count", total);
        emit(allTrue, message, "sometimes_all", true, d, 2);
    }

    private static void emit(boolean condition, String message, String assertType, boolean mustHit,
                              Map<String, Object> details, int skipFrames) {
        StackTraceElement[] stack = new Throwable().getStackTrace();
        StackTraceElement caller = stack.length > skipFrames ? stack[skipFrames] : stack[stack.length - 1];
        String file = caller.getFileName() != null ? caller.getFileName() : "unknown";
        int line = caller.getLineNumber();
        String id = callsiteId(file, line);

        String declKey = assertType + ":" + message;
        if (declared.add(declKey)) {
            SdkWriter.emit(buildJson(false, false, message, assertType, mustHit, id, null, file, line));
        }
        SdkWriter.emit(buildJson(true, condition, message, assertType, mustHit, id, details, file, line));
    }

    private static String callsiteId(String file, int line) {
        try {
            byte[] input = (file + ":" + line).getBytes(java.nio.charset.StandardCharsets.UTF_8);
            java.security.MessageDigest md = java.security.MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(input);
            return String.format("%02x%02x%02x%02x", hash[0], hash[1], hash[2], hash[3]);
        } catch (java.security.NoSuchAlgorithmException e) {
            return String.format("%08x", ((long) (file + ":" + line).hashCode()) & 0xFFFFFFFFL);
        }
    }

    private static String buildJson(boolean hit, boolean condition, String message, String assertType,
                                    boolean mustHit, String id, Map<String, Object> details,
                                    String file, int line) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"openthesis_assert\":{");
        sb.append("\"hit\":").append(hit).append(",");
        sb.append("\"condition\":").append(condition).append(",");
        sb.append("\"message\":").append(jsonStr(message)).append(",");
        sb.append("\"assert_type\":").append(jsonStr(assertType)).append(",");
        sb.append("\"must_hit\":").append(mustHit).append(",");
        sb.append("\"id\":").append(jsonStr(id));
        if (details != null && !details.isEmpty()) {
            sb.append(",\"details\":").append(mapToJson(details));
        }
        sb.append(",\"location\":{\"file\":").append(jsonStr(file)).append(",\"line\":").append(line).append("}");
        sb.append("}}");
        return sb.toString();
    }

    static String jsonStr(String s) {
        if (s == null) return "null";
        StringBuilder sb = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\n': sb.append("\\n");  break;
                case '\r': sb.append("\\r");  break;
                case '\t': sb.append("\\t");  break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append("\"");
        return sb.toString();
    }

    static String mapToJson(Map<String, Object> m) {
        if (m == null || m.isEmpty()) return "{}";
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : m.entrySet()) {
            if (!first) sb.append(",");
            first = false;
            sb.append(jsonStr(entry.getKey())).append(":").append(valueToJson(entry.getValue()));
        }
        sb.append("}");
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    static String valueToJson(Object v) {
        if (v == null) return "null";
        if (v instanceof Boolean || v instanceof Number) return v.toString();
        if (v instanceof String) return jsonStr((String) v);
        if (v instanceof Map) {
            try {
                return mapToJson((Map<String, Object>) v);
            } catch (ClassCastException e) {
                return jsonStr(v.toString());
            }
        }
        return jsonStr(v.toString());
    }
}

File: sdk/rust/src/assert.rs
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::sync::{Mutex, OnceLock};

static WRITER: OnceLock<Mutex<Option<File>>> = OnceLock::new();
static DECLARED: OnceLock<Mutex<HashMap<String, bool>>> = OnceLock::new();

// ever_since armed state: key → still_holding
static EVER_SINCE: OnceLock<Mutex<HashMap<String, bool>>> = OnceLock::new();

fn writer() -> &'static Mutex<Option<File>> {
    WRITER.get_or_init(|| {
        let file = open_output_file();
        Mutex::new(file)
    })
}

fn declared() -> &'static Mutex<HashMap<String, bool>> {
    DECLARED.get_or_init(|| Mutex::new(HashMap::new()))
}

fn ever_since_state() -> &'static Mutex<HashMap<String, bool>> {
    EVER_SINCE.get_or_init(|| Mutex::new(HashMap::new()))
}

fn open_output_file() -> Option<File> {
    let dir = std::env::var("OPENTHESIS_OUTPUT_DIR")
        .ok()
        .filter(|s| !s.is_empty())
        .or_else(|| {
            std::env::var("OPENTHESIS_SDK_LOCAL_OUTPUT")
                .ok()
                .filter(|s| !s.is_empty())
        })?;

    let path = if dir.ends_with('/') {
        format!("{}sdk.jsonl", dir)
    } else {
        format!("{}/sdk.jsonl", dir)
    };

    OpenOptions::new()
        .append(true)
        .create(true)
        .open(&path)
        .ok()
}

fn callsite_id(file: &str, line: u32) -> String {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    let mut h = DefaultHasher::new();
    format!("{}:{}", file, line).hash(&mut h);
    format!("{:08x}", h.finish() & 0xFFFFFFFF)
}

fn write_json(val: &Value) {
    let mut guard = writer().lock().unwrap_or_else(|e| e.into_inner());
    if let Some(ref mut f) = *guard {
        let mut bytes = serde_json::to_vec(val).unwrap_or_default();
        bytes.push(b'\n');
        let _ = f.write_all(&bytes);
    }
}

pub fn emit(
    condition: bool,
    message: &str,
    assert_type: &str,
    must_hit: bool,
    details: Option<Value>,
    file: &str,
    line: u32,
) {
    let id = callsite_id(file, line);
    let decl_key = format!("{}:{}", assert_type, message);

    // Emit declaration on first encounter.
    {
        let mut decl = declared().lock().unwrap_or_else(|e| e.into_inner());
        if !decl.contains_key(&decl_key) {
            decl.insert(decl_key.clone(), true);
            let decl_val = json!({
                "openthesis_assert": {
                    "hit": false,
                    "condition": false,
                    "message": message,
                    "assert_type": assert_type,
                    "must_hit": must_hit,
                    "id": id,
                    "location": { "file": file, "line": line }
                }
            });
            write_json(&decl_val);
        }
    }

    let mut body = json!({
        "openthesis_assert": {
            "hit": true,
            "condition": condition,
            "message": message,
            "assert_type": assert_type,
            "must_hit": must_hit,
            "id": id,
            "location": { "file": file, "line": line }
        }
    });

    if let Some(d) = details {
        body["openthesis_assert"]["details"] = d;
    }

    write_json(&body);
}

// ever_since: armed on first call; once condition becomes false, stays false.
pub fn emit_ever_since(
    condition: bool,
    message: &str,
    details: Option<Value>,
    file: &str,
    line: u32,
) {
    let key = format!("{}:{}", file, line);
    let mut state = ever_since_state().lock().unwrap_or_else(|e| e.into_inner());
    let holding = state.entry(key).or_insert(true);
    if *holding && !condition {
        *holding = false;
    }
    let effective = *holding;
    drop(state);

    emit(effective, message, "always", true, details, file, line);
}

/// Assert that `condition` is true every time this is called.
/// A single false evaluation is a violation.
#[macro_export]
macro_rules! always {
    ($condition:expr, $message:expr) => {
        $crate::assert::emit($condition, $message, "always", true, None, file!(), line!())
    };
    ($condition:expr, $message:expr, $details:expr) => {
        $crate::assert::emit(
            $condition,
            $message,
            "always",
            true,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that `condition` is true whenever this code is reached,
/// but also passes if never reached.
#[macro_export]
macro_rules! always_or_unreachable {
    ($condition:expr, $message:expr) => {
        $crate::assert::emit(
            $condition,
            $message,
            "always_or_unreachable",
            false,
            None,
            file!(),
            line!(),
        )
    };
    ($condition:expr, $message:expr, $details:expr) => {
        $crate::assert::emit(
            $condition,
            $message,
            "always_or_unreachable",
            false,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that `condition` is true at least once across the entire run.
#[macro_export]
macro_rules! sometimes {
    ($condition:expr, $message:expr) => {
        $crate::assert::emit($condition, $message, "sometimes", true, None, file!(), line!())
    };
    ($condition:expr, $message:expr, $details:expr) => {
        $crate::assert::emit(
            $condition,
            $message,
            "sometimes",
            true,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that this code path is reached at least once.
#[macro_export]
macro_rules! reachable {
    ($message:expr) => {
        $crate::assert::emit(true, $message, "reachable", true, None, file!(), line!())
    };
    ($message:expr, $details:expr) => {
        $crate::assert::emit(
            true,
            $message,
            "reachable",
            true,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that this code path is never reached.
#[macro_export]
macro_rules! unreachable_assert {
    ($message:expr) => {
        $crate::assert::emit(false, $message, "unreachable", false, None, file!(), line!())
    };
    ($message:expr, $details:expr) => {
        $crate::assert::emit(
            false,
            $message,
            "unreachable",
            false,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that once `condition` becomes true it stays true for the rest of the run.
/// Tracks armed state per call site.
#[macro_export]
macro_rules! ever_since {
    ($condition:expr, $message:expr) => {
        $crate::assert::emit_ever_since($condition, $message, None, file!(), line!())
    };
    ($condition:expr, $message:expr, $details:expr) => {
        $crate::assert::emit_ever_since($condition, $message, Some($details), file!(), line!())
    };
}

/// Assert that `left > right` every time called.
/// The violation details include both operand values.
#[macro_export]
macro_rules! always_greater_than {
    ($left:expr, $right:expr, $message:expr) => {{
        let l = $left;
        let r = $right;
        $crate::assert::emit(
            l > r,
            $message,
            "always",
            true,
            Some(serde_json::json!({"left": l, "right": r, "op": ">"})),
            file!(),
            line!(),
        )
    }};
}

/// Assert that `left > right` at least once across the run.
#[macro_export]
macro_rules! sometimes_greater_than {
    ($left:expr, $right:expr, $message:expr) => {{
        let l = $left;
        let r = $right;
        $crate::assert::emit(
            l > r,
            $message,
            "sometimes",
            true,
            Some(serde_json::json!({"left": l, "right": r, "op": ">"})),
            file!(),
            line!(),
        )
    }};
}

/// Assert that `left == right` every time called.
#[macro_export]
macro_rules! always_equal {
    ($left:expr, $right:expr, $message:expr) => {{
        let l = $left;
        let r = $right;
        $crate::assert::emit(
            l == r,
            $message,
            "always",
            true,
            Some(serde_json::json!({"left": l, "right": r, "op": "=="})),
            file!(),
            line!(),
        )
    }};
}

/// Assert that `left == right` at least once across the run.
#[macro_export]
macro_rules! sometimes_equal {
    ($left:expr, $right:expr, $message:expr) => {{
        let l = $left;
        let r = $right;
        $crate::assert::emit(
            l == r,
            $message,
            "sometimes",
            true,
            Some(serde_json::json!({"left": l, "right": r, "op": "=="})),
            file!(),
            line!(),
        )
    }};
}

/// Assert that each distinct key under label is observed at least once.
/// Call with the same label and a different key each time.
#[macro_export]
macro_rules! sometimes_each {
    ($label:expr, $key:expr) => {
        $crate::assert::emit(
            true,
            &format!("{}:{}", $label, $key),
            "sometimes",
            true,
            None,
            file!(),
            line!(),
        )
    };
    ($label:expr, $key:expr, $details:expr) => {
        $crate::assert::emit(
            true,
            &format!("{}:{}", $label, $key),
            "sometimes",
            true,
            Some($details),
            file!(),
            line!(),
        )
    };
}

/// Assert that all named conditions are simultaneously true at least once.
/// Pass a HashMap<&str, bool> as the named conditions.
#[macro_export]
macro_rules! sometimes_all {
    ($message:expr, $named_bools:expr) => {{
        let satisfied: usize = $named_bools.values().filter(|&&v| v).count();
        let total: usize = $named_bools.len();
        let details = serde_json::json!({
            "satisfied_count": satisfied,
            "total_count": total,
            "sub_goals": $named_bools,
        });
        $crate::assert::emit(
            satisfied == total,
            $message,
            "sometimes_all",
            true,
            Some(details),
            file!(),
            line!(),
        )
    }};
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_emit_no_output_dir() {
        // Should not panic when OPENTHESIS_OUTPUT_DIR is not set.
        emit(true, "test message", "always", true, None, "test.rs", 1);
    }

    #[test]
    fn test_emit_with_details() {
        emit(
            false,
            "violation message",
            "always",
            true,
            Some(json!({"key": "value"})),
            "test.rs",
            2,
        );
    }

    #[test]
    fn test_ever_since_basic() {
        emit_ever_since(true, "ever_since test", None, "test.rs", 100);
        emit_ever_since(false, "ever_since test", None, "test.rs", 100);
        // After going false, subsequent calls should remain false.
        emit_ever_since(true, "ever_since test", None, "test.rs", 100);
    }

    #[test]
    fn test_callsite_id_stable() {
        let id1 = callsite_id("foo.rs", 42);
        let id2 = callsite_id("foo.rs", 42);
        assert_eq!(id1, id2);
        let id3 = callsite_id("foo.rs", 43);
        assert_ne!(id1, id3);
    }
}
